### PR TITLE
Trie-based approach for global ignores

### PR DESCRIPTION
**Pull request by mcculls**
This PR replaces the hand-crafted ignores filter with a string-based trie, generated from a list of package and class names.

Most of the lines are associated with one of these numbers:
```
# 0 = global allows
# 1 = system-level ignores
# 2 = additional ignores
```
For example:
```
# ...
1 java.
# allow exception profiling instrumentation
0 java.lang.Throwable
0 java.net.HttpURLConnection
0 java.net.URL
0 java.rmi.
0 java.util.concurrent.
0 java.util.logging.
# Concurrent instrumentation modifies the structure of the Cleaner class incompatibly with java9+ modules.
1 java.util.logging.LogManager$Cleaner
# ...
2 com.codahale.metrics.
# We instrument servlets
0 com.codahale.metrics.servlets.
# ...
```
There are also a few special cases for when the initial prefix match needs to be followed by a specific suffix match:
```
# 3 = ignore if endsWith Proxy
# 4 = allow if endsWith HttpMessageConverter

3 com.mchange.v2.c3p0.
4 org.springframework.http.converter.
```

Lines that end in `.` or `$` are considered as wildcards that will match classes who share that prefix, unless another longer line matches the class.  This lets us override the filter for specific classes or sub-packages.

A gradle task takes the `.trie` file and generates a Java class embedding the compressed string trie as a constant. The source is generated to the standard source directory so it can be checked into CI. The gradle task supports incremental builds and will only update the generated file if the `.trie` file (or the generator itself) changes.
 
Performance results:

I tested the old and new approaches using a set of almost 450,000 class names collected from common IDEs. Each "operation" in the performance test applies the matcher to all 450,000 class names.  Replacing the old `startsWith` matching with a trie is 2.5 times faster. This is because even with the trie the global ignores matcher still does some `contains` matching, to exclude certain proxies.

When I tested the trie on its own without these contains checks it was ~9 times faster than the old approach.

Java 8:
```
Benchmark                          Mode  Cnt   Score   Error  Units
ClassNameParsing.originalMatcher  thrpt   25   5.149 ± 0.032  ops/s
ClassNameParsing.trieMatcher      thrpt   25  13.141 ± 0.046  ops/s
ClassNameParsing.trie             thrpt   25  47.790 ± 1.218  ops/s
```
Java 11:
```
Benchmark                          Mode  Cnt   Score   Error  Units
ClassNameParsing.originalMatcher  thrpt   25   6.499 ± 0.121  ops/s
ClassNameParsing.trieMatcher      thrpt   25  16.058 ± 0.170  ops/s
ClassNameParsing.trie             thrpt   25  41.057 ± 2.865  ops/s
```

The list of package/class names should be easier to manage than the previous hand-crafted approach while still supporting the split between system and additional ignores.

The new trie also supports matching against internal class names, such as `java/lang/Throwable`, which means we could consider moving some ignores filtering to the `ClassFileTransformer` where the JVM passes in internal class names.

### CODE REVIEW

File: internal-api/src/main/java/datadog/trace/util/ClassNameTrie.java [hunk 1]
```diff
@@ -0,0 +1,523 @@
...
+    int dataIndex = 0;
+    int result = -1;
+
+    while (keyIndex < keyLength) {
+      char c = key.charAt(keyIndex++);
+      char branchCount = data[dataIndex++];
+
+      // trie is ordered, so we can use binary search to pick the right branch
+      int branchIndex =
+          Arrays.binarySearch(data, dataIndex, dataIndex + branchCount, c == '/' ? '.' : c);
```
Comment (bantonsson) [inline internal-api/src/main/java/datadog/trace/util/ClassNameTrie.java:123]: Nice to treat `/` and `.` the same...

File: internal-api/src/main/java/datadog/trace/util/ClassNameTrie.java [hunk 2]
```diff
@@ -0,0 +1,523 @@
...
+ *               | 1 | c | length (0)
+ *               | 2 | l | o | length (6) | length (27) | jump (40)
+ *               assic. | 2 | L | s | length (5) | length (14) | jump (6)
+ *                      ogger | leaf (0)
+ *                      pi.LoggingEven | 1 | t | bud (0)
+ *                                     | 1 | V | length (1)
+ *                                     O | leaf (0)
+ *               re.AsyncAppenderBase$Worker | leaf(0)
+ * </pre>
+ */
```
Comment (bantonsson) [inline internal-api/src/main/java/datadog/trace/util/ClassNameTrie.java:87]: Great comment!

File: dd-java-agent/agent-tooling/src/main/java/datadog/trace/agent/tooling/bytebuddy/matcher/IgnoredClassNameTrie.java [hunk 3]
```diff
@@ -0,0 +1,170 @@
+package datadog.trace.agent.tooling.bytebuddy.matcher;
+
+import datadog.trace.util.ClassNameTrie;
+
+// Generated from 'ignored_class_name.trie' - DO NOT EDIT!
```
Comment (bantonsson) [inline dd-java-agent/agent-tooling/src/main/java/datadog/trace/agent/tooling/bytebuddy/matcher/IgnoredClassNameTrie.java:5]: Do we really need to check these in if they are generated by the build?

**Review by bantonsson — APPROVED** [review]
👍

File: dd-java-agent/agent-tooling/src/main/java/datadog/trace/agent/tooling/bytebuddy/matcher/IgnoredClassNameTrie.java (hunk 3 above)
Comment (mcculls) [reply dd-java-agent/agent-tooling/src/main/java/datadog/trace/agent/tooling/bytebuddy/matcher/IgnoredClassNameTrie.java:5]: We don't have to, we could put them under a generated source directory

Comment (bantonsson) [reply dd-java-agent/agent-tooling/src/main/java/datadog/trace/agent/tooling/bytebuddy/matcher/IgnoredClassNameTrie.java:5]: I think that would be great

**Pull request merged**